In [1]:
%load_ext autoreload
%autoreload 2
%matplotlib inline
from my_modules import risk
from my_modules import data
from my_modules import portfolio
from my_modules import index
from my_modules import model

import numpy as np
import pandas as pd

## Present Value of Liabilities and Funding Ratio

In [3]:
def discount(t, interest_rate):
    """
    Compute the price of a pure discount bond that pays a dollar at time t, given interest rate r.
    Assumes that the yield curve is flat (duration does not matter)
    """
    return (1 + interest_rate)**-t # 1 over (1 + r)^t

In [4]:
discount(10,0.03) #74 cents to get back a dollar in 10 years

0.7440939148967249

In [5]:
0.7440939148967249 * (1.03)**10

1.0

In [7]:
def pv(l,r):
    """
    Compute the present value of a sequence of liabilities
    l is indexed by time, and values are the amounts of each liability
    returns the present value of the sequence
    """
    dates = l.index
    discounts = discount(dates,r)
    return (discounts * l).sum()

In [8]:
liabilities = pd.Series(data=[1,1.5,2,2.5], index=[3,3.5,4,4.5]) # Payment schedules, e.g in 3 years need to pay 1MM etc
liabilities

3.0    1.0
3.5    1.5
4.0    2.0
4.5    2.5
dtype: float64

In [9]:
pv(liabilities,0.03)

6.233320315080045

_6.23 indicates that the 7MM in liability has a present value of 6.23 because of the time value of money i.e. will be able to meet future liabilities with the 6.23 today_

In [18]:
def funding_ratio(assets, liabilities, r):
    """
    Computes funding ratio given a set of assets, liabilites and a constant insterest rate
    """
    return assets/pv(liabilities, r)

In [19]:
funding_ratio(5, liabilities, 0.03)

0.8021407126958777

.802 means that there is only 80% funding. It is not possible to buy bonds in order to pay off the liabilites

In [15]:
funding_ratio(5, liabilities, 0.02) # Decreasing interests rates is obviously bad

0.7720304366941648

In [16]:
import ipywidgets as widgets
from IPython.display import display
%matplotlib inline

In [21]:
def show_funding_ratio(assets, r):
    fr = funding_ratio(assets, liabilities, r)
    print(f'{fr*100:.2f}')
    
controls = widgets.interactive(show_funding_ratio,
                              assets= widgets.IntSlider(min=1,max=15,step=1, value=5),
                              r=(0,.2,.01))

display(controls)

interactive(children=(IntSlider(value=5, description='assets', max=15, min=1), FloatSlider(value=0.1, descript…